translator/trace/jaeger/traces_to_jaegerproto.go

```diff
@@ -0,0 +1,361 @@
+	}
+	if traceIDLow == 0 && traceIDHigh == 0 {
+		return model.TraceID{}, errZeroTraceID
+	}
+	return model.TraceID{
+		Low:  traceIDLow,
+		High: traceIDHigh,
+	}, nil
+}
+
+func spanIDToJaegerProto(spanID pdata.SpanID) (model.SpanID, error) {
+	uSpanID, err := tracetranslator.BytesToUInt64SpanID([]byte(spanID))
+	if err != nil {
+		return model.SpanID(0), err
+	}
+	if uSpanID == 0 {
+		return model.SpanID(0), errZeroSpanID
+	}
+	return model.SpanID(uSpanID), nil
+}
+
+// makeJaegerProtoReferences constructs jaeger span references based on parent span ID and span links
+func makeJaegerProtoReferences(
+	links pdata.SpanLinkSlice,
+	parentSpanID pdata.SpanID,
+	traceID model.TraceID,
+) ([]model.SpanRef, error) {
+	parentSpanIDSet := len(parentSpanID.Bytes()) != 0
+	if !parentSpanIDSet && links.Len() == 0 {
+		return nil, nil
+	}
+
+	refsCount := links.Len()
+	if parentSpanIDSet {
+		refsCount++
+	}
+
+	refs := make([]model.SpanRef, 0, refsCount)
+
+	// Put parent span ID at the first place because usually backends look for it
+	// as the first CHILD_OF item in the model.SpanRef slice.
+	if parentSpanIDSet {
+		jParentSpanID, err := spanIDToJaegerProto(parentSpanID)
+		if err != nil {
+			return nil, fmt.Errorf("OC incorrect parent span ID: %v", err)
+		}
+
+		refs = append(refs, model.SpanRef{
+			TraceID: traceID,
+			SpanID:  jParentSpanID,
+			RefType: model.SpanRefType_CHILD_OF,
+		})
+	}
+
+	for i := 0; i < links.Len(); i++ {
+		link := links.At(i)
+		if link.IsNil() {
+			continue
+		}
+
+		traceID, err := traceIDToJaegerProto(link.TraceID())
+		if err != nil {
+			continue // skip invalid link
+		}
+
+		spanID, err := spanIDToJaegerProto(link.SpanID())
+		if err != nil {
+			continue // skip invalid link
+		}
+
+		refs = append(refs, model.SpanRef{
+			TraceID: traceID,
+			SpanID:  spanID,
+
+			// Since Jaeger RefType is not captured in internal data,
+			// use SpanRefType_FOLLOWS_FROM by default.
+			// SpanRefType_CHILD_OF supposed to be set only from parentSpanID.
+			RefType: model.SpanRefType_FOLLOWS_FROM,
+		})
+	}
+
+	return refs, nil
+}
+
+func spanEventsToJaegerProtoLogs(events pdata.SpanEventSlice) []model.Log {
+	if events.Len() == 0 {
+		return nil
+	}
+
+	logs := make([]model.Log, 0, events.Len())
+	for i := 0; i < events.Len(); i++ {
+		event := events.At(i)
+		if event.IsNil() {
+			continue
+		}
+
+		logs = append(logs, model.Log{
+			Timestamp: internal.UnixNanoToTime(event.Timestamp()),
+			Fields:    attributesToJaegerProtoTags(event.Attributes()),
+		})
+	}
+
+	return logs
+}
+
+func appendTagFromSpanKind(tags []model.KeyValue, spanKind pdata.SpanKind) []model.KeyValue {
+	tag := model.KeyValue{
+		Key:   tracetranslator.TagSpanKind,
+		VType: model.ValueType_STRING,
+	}
+
+	switch spanKind {
+	case pdata.SpanKindCLIENT:
+		tag.VStr = "client"
+	case pdata.SpanKindSERVER:
+		tag.VStr = "server"
+	case pdata.SpanKindPRODUCER:
+		tag.VStr = "producer"
+	case pdata.SpanKindCONSUMER:
+		tag.VStr = "consumer"
```

Comment:
We have these string costs defined near `OpenTracingSpanKind` in `protospan_translation.go`.

Reply:
changed